File: 067.Add Binary 二进制数相加/067.Add Binary 二进制数相加.py
#------------------------------第一种------------------------------------
class Solution:
    # @param a, a string
    # @param b, a string
    # @return a string
    def addBinary(self, a, b):
        result, carry, val = "", 0, 0
        for i in xrange(max(len(a), len(b))):
            val = carry
            if i < len(a):
                val += int(a[-(i + 1)])
            if i < len(b):
                val += int(b[-(i + 1)])
            carry, val = val / 2, val % 2
            result += str(val)
        if carry:
            result += str(carry)
        return result[::-1]
#------------------------------第二种------------------------------------
class Solution:
     def addBinary(self, a, b):
         if len(a)==0: return b
         if len(b)==0: return a
         if a[-1] == '1' and b[-1] == '1':
             return self.addBinary(self.addBinary(a[0:-1],b[0:-1]),'1')+'0'
         if a[-1] == '0' and b[-1] == '0':
             return self.addBinary(a[0:-1],b[0:-1])+'0'
         else:
             return self.addBinary(a[0:-1],b[0:-1])+'1'　
#----------------------------第三种---------------------------------------
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        res = ''
        carry = 0
        i, j = len(a) - 1, len(b) - 1
        while i >= 0 or j >= 0:
            su = carry
            if i >= 0:
                su += int(a[i])
            if j >= 0:
                su += int(b[j]) 
            carry = su / 2
            res += str(su % 2)
            i -= 1
            j -= 1
  
        if carry > 0:
            res += str(carry)
  
        return res[::-1] 　　

File: 067.Add Binary 二进制数相加/067.Add Binary 二进制数相加.java
//------------------------------第一种------------------------------------
public class Solution {
    public String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int i = a.length() - 1, j = b.length() -1, carry = 0;
        while (i >= 0 || j >= 0) {
            int sum = carry;
            if (j >= 0) sum += b.charAt(j--) - '0';
            if (i >= 0) sum += a.charAt(i--) - '0';
            sb.append(sum % 2);
            carry = sum / 2;
        }
        if (carry != 0) sb.append(carry);
        return sb.reverse().toString();
    }
}　　
//------------------------------第二种------------------------------------
public class AddBinary {   
    public String addBinary(String a, String b) {
        StringBuilder result = new StringBuilder();
        int pointerA = a.length()-1;
        int pointerB = b.length()-1;
        int carry = 0;
        while(pointerA>=0 || pointerB>=0){
            int sum = carry;
            if(pointerA>=0){
                sum += (a.charAt(pointerA)-'0');
                pointerA--;
            }
            if(pointerB>=0){
                sum += (b.charAt(pointerB)-'0');
                pointerB--;
            }
            result.append(sum%2);
            carry = sum/2;
        }
        if(carry!=0){
            result.append('1');
        }
        return result.reverse().toString();
    }
}